climada/util/coordinates.py

```diff
@@ -1554,21 +1569,32 @@ def read_raster(file_name, band=None, src_crs=None, window=None, geometry=None,
 
     return dst_meta, intensity.reshape(dst_shape)
 
-def read_raster_bounds(path, bounds, res=None, bands=None):
+def read_raster_bounds(path, bounds, res=None, bands=None, resampling="nearest",
+                       global_origin=None):
     """Read raster file within given bounds and refine to given resolution
 
-    Makes sure that the extent of pixel centers covers the specified regions
+    The procedure makes sure that the extent of pixel centers covers the specified region.
```

Comment:
I think the idea is that sometimes this will return cells that extend beyond the bounds, but agreed, there might be a clearer way to say this?

Reply:
When you want to read a raster file within given bounds, it's not entirely clear how that is even defined:

![image](https://user-images.githubusercontent.com/57705593/148514287-0d010299-d480-4af8-aeba-251e88c923dd.png)

The window reading functionality of rasterio, for example, always assumes the left interpretation, where only those cells are selected for which the cell centers (black dots) lie within the selected bounds (blue rectangle).

However, in many applications, for example when you apply bilinear interpolation to evaluate your raster at other positions than just the cell centers, the right interpretation is required in order to be able to evaluate the grid everywhere within the blue rectangle up to its boundary. This is because bilinear interpolation requires all four neighboring cell centers of a point.

A user who wants to have the seconds interpretation could just enlarge the bounds manually and apply a function that uses the first interpretation. However, before actually opening the raster file, the user doesn't know how much to enlarge the bounds. This is why it's convenient that this function directly offers to apply the interpretation on the right.

By the way, the function returns not only the array, but the transform together with the array to make sure that the user knows which array values actually correspond to which geographical position.

Reply:
Aaa thanks, so it is "just" about the padding?

Reply:
Since you are more familiar with that term, I decided to transform this feature into a padding-related kwarg called `pad_cells`:
https://github.com/CLIMADA-project/climada_python/blob/b74a06446e153b36dea8ab5f948002858160b168/climada/util/coordinates.py#L1571-L1604

Reply:
Is there a reason `pad_cells` is a float? An integer makes more intuitive sense, and is less likely to be confused by the user for a padding measured in degrees.

Reply:
Also, shouldn't the padding be applied at the resolution of the source data, not the destination? (If I've understood this right)

Reply:
There is no reason for `pad_cells` to be an integer. It is a conventional padding parameter measured in units of "cell width". If it would want to force it to an integer, we would need to artificially check whether it's an integer and raise an exception even though the procedure would work perfectly fine with floats.

As for the source/destination question: The pad size is in terms of the cells of the destination grid (which might be the same as the source grid if neither `res` or `global_origin` are specified by the user). I clarified this in the updated docstrings:
https://github.com/CLIMADA-project/climada_python/blob/ca2ba7dcc144f2c3bbcad4c9c60043c32322fd8b/climada/util/coordinates.py#L1601-L1605